Please add FreeSSL.tech in-browser ACME V2 client

SpeedUpWebsite opened the pull request:
Please add FreeSSL.tech in-browser ACME V2 client. Tested with the live API. Certificate issue, revoke and LE account key change implemented. Thank you.

SpeedUpWebsite commented:
FYI, I have followed following best-practices:

- I have kept the domain private key + CSR generation saparate and optional from the SSL cert generation process. 

- The CSR is generated at a different page (https://freessl.tech/csr-ssl-certificate-signing-request/). User need to manually copy CSR from that page to the SSL cert generation form https://freessl.tech/ap/certificate-create

- User can download the CSR generation app from github and run locally https://github.com/SpeedUpWebsite/san-csr 

- In the SSL cert generation page I have written "If your web hosting control panel has an option to generate CSR, we highly recommend using that tool." in bold.

Let's Encrypt guidelines for web/browser based ACME clients still not published and I have worked months on this client. If you could have publish the guidelines earlier, that could save my time.

cpu commented:
I understand that this is a frustrating experience @SpeedUpWebsite. I appreciate that you have put a lot of time and care into your client. I'm going to bump this conversation internally to try and reach consensus on how we plan to address browser based clients and the ACME client list.

tdelmas commented:
(I think it's not the best place to say it, but as It concerns that service and it may impact the conversation that Let's Encrypt will have about web based csr generation, here are my remarks about that service)

If Let's Encrypt thinks it's acceptable to generate on a webpage the csr, I've a 

few remarks about your tool:

- It (tries to) loads javascript from insecure source (http://freegoeip.net)

- It loads javascript from third-party domains:
  - twitter.com
  - facebook.com

- MitM prevention:
  - The http http://freessl.tech/csr-ssl-certificate-signing-request/ doesn't redirects to https://freessl.tech/csr-ssl-certificate-signing-request/
  - It doesn't use the HSTS header
  - freessl.tech isn't in the HSTS preload list: https://hstspreload.org/?domain=freessl.tech

- https configuration needs improvements: 
  - https://www.ssllabs.com/ssltest/analyze?d=freessl.tech (grade B - below A is not acceptable for a security product)
  - https://tls.imirhil.fr/https/freessl.tech (score C - but that notation is harsh, B may be acceptable)
  - https://www.htbridge.com/ssl/?id=0t8WdfgX (grade B)


- SRI (https://developer.mozilla.org/fr/docs/Web/Security/Subresource_Integrity)
  - No SRI for external javascript 
  - No SRI for internal javascript

- Security headers that may be relevant:
  - No Content-Security-Policy
  - No X-XSS-Protection
  - No X-Content-Type-Options 
  Details: https://observatory.mozilla.org/analyze/freessl.tech and https://securityheaders.com/?followRedirects=on&hide=on&q=freessl.tech

cpu commented:
Hi again @SpeedUpWebsite,

Unfortunately I'm going to have to close this PR. After discussing the policy question internally we've decided that browser based clients are too difficult for us to evaluate reliably, and more importantly, do not support the renewal automation that we think is critical to the service. On this basis we've [updated the client options page](https://github.com/letsencrypt/website/pull/377) to describe that we won't be listing browser based ACME clients on the website.

Thanks for understanding,

SpeedUpWebsite commented:
@cpu No problem. I am developing another client, in PHP. It is a server-side client and automates the process. I'll create PR for that as soon I complete.
And I'll remove this in-browser client from https://freessl.tech. Though I may publish that in a different domain (I'll not create PR for it again, of course). Actually, I developed the in-browser client so that the less tech-savvy people can test drive with it first. In the final step (3 out of 3) page, I'll put a link to the PHP client and let users know that the PHP client automates everything.

@tdelmas Thank you very much for pointing out the issues of https://freessl.tech. I have fixed some and will fix remaining. Here is the detail:
-  http://freessl.tech/* now redirects to https://freessl.tech/*
-  I have implemented HSTS header.
-  https://hstspreload.org/?domain=freessl.tech is now green and I have submitted to the HSTS preload list.
-  Grade A+ at https://www.ssllabs.com/ssltest/analyze.html?d=freessl.tech&hideResults=on
-  Score A+ at https://tls.imirhil.fr/https/freessl.tech
-  Grade A+ at https://www.htbridge.com/ssl/?id=0t8WdfgX
-  I have implemented Content-Security-Policy
-  X-XSS-Protection has been done
-  X-Content-Type-Options has been done
-  Grade A+ at https://securityheaders.com/?followRedirects=on&hide=on&q=freessl.tech
-  Grade B at https://observatory.mozilla.org/analyze/freessl.tech  (It was F initially).

tdelmas commented:
> Thank you very much for pointing out the issues of https://freessl.tech. I have fixed some and will fix remaining. Here is the detail

@SpeedUpWebsite I'm glad to see these changes! (My main remaining concern are the number of javascript requests to external services on the webpage that still generate the private key of the ACME account)

SpeedUpWebsite commented:
@cpu My pleasure.

SpeedUpWebsite commented:
@tdelmas I have fixed the JS issues.

- Removed using http://freegoeip.net (this API is expired)
- Stopped using JS from third-party domains like twitter.com, facebook.com. Now the site using internal JS for the respective functionalities.
- Implemented SRI for internal javascript and CSS.

Thank you again for your feedback.
